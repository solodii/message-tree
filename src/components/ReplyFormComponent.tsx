import * as React from 'react';
import { Draft } from '../state/types';

export interface Props {
  subjectId: string;
  draft: Readonly<Draft>;
  isValid: boolean;
  onChange: (value: string, subjectId: string) => void;
  onCancel: (subjectId: string) => void;
  onSubmit: (subjectId: string) => void;
}

export default class ReplyFormComponent extends React.Component<Props, {}> {
  constructor(props: Props) {
    super(props);
    this.onInputChange = this.onInputChange.bind(this);
    this.onCancel = this.onCancel.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  public render(): JSX.Element {
    const {
      draft: { text },
      isValid
    } = this.props;
    return (
      <div className="form">
        <div>
          <textarea
            className="input"
            autoFocus={true}
            placeholder="Type your message here..."
            value={text}
            onChange={this.onInputChange}
          />
        </div>
        <div>
          <button
            className="cancel"
            onClick={this.onCancel}
          >
            Cancel
          </button>
          <button
            className="submit"
            disabled={!isValid}
            onClick={this.onSubmit}
          >
            Submit
          </button>
        </div>
      </div>
    );
  }

  private onInputChange(e: React.ChangeEvent<HTMLTextAreaElement>): void {
    this.props.onChange(e.target.value, this.props.subjectId);
  }

  private onCancel(): void {
    this.props.onCancel(this.props.subjectId);
  }

  private onSubmit(): void {
    this.props.onSubmit(this.props.subjectId);
  }
}
