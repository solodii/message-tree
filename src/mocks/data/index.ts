export * from './constants';
export {
  drafts,
  draftsBySubject,
  invalidDraft,
  validDraft
} from './drafts';
export {
  messages,
  messagesById,
  messagesBySubject,
  messagesWithDraft,
  messagesWithoutDraft,
  messageTrees
} from './messages';
export { state } from './state';
