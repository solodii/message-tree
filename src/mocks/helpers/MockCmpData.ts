import { ShallowWrapper } from 'enzyme';

export interface MockCmpData<P, S = {}> {
  wrapper: ShallowWrapper<P, S>;
  initialProps: P;
}
