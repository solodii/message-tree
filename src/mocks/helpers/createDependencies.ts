import { Subject } from 'rxjs';
import { Dependencies } from '../../epics';

export function createDependencies(): Dependencies {
  const dependencies = { socket: new Subject() };
  return dependencies as Dependencies;
}
