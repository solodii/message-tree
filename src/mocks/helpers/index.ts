export { createMiddlewareAPI } from './createMiddlewareAPI';
export { createDependencies } from './createDependencies';
export { MockCmpData } from './MockCmpData';
