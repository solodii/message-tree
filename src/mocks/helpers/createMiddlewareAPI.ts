import { MiddlewareAPI } from 'redux';
import { State } from '../../state/types';
import { state } from '../data/state';

export function createMiddlewareAPI(): MiddlewareAPI<State> {
  return {
    dispatch: jest.fn(),
    getState: jest.fn(() => state)
  };
}
