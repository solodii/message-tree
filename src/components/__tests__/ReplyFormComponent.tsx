import * as React from 'react';
import { shallow } from 'enzyme';
import {
  invalidDraft,
  validDraft
} from '../../mocks/data';
import { MockCmpData } from '../../mocks/helpers';
import ReplyFormComponent, { Props } from '../ReplyFormComponent';

describe('ReplyFormComponent', () => {
  const validMockData = createCmpWithValidDraft();
  const invalidMockData = createCmpWithInvalidDraft();

  describe('VALID_DRAFT_SUBJECT', () => {
    test('renders correctly', () => {
      expect(validMockData.wrapper).toMatchSnapshot();
    });
  });

  describe('INVALID_DRAFT_SUBJECT', () => {
    test('renders correctly', () => {
      expect(invalidMockData.wrapper).toMatchSnapshot();
    });
  });

  describe('callbacks', () => {
    const {
      initialProps: {
        subjectId,
        onChange,
        onCancel,
        onSubmit
      },
      wrapper
    } = validMockData;

    test('.input onChange is called after input change', () => {
      const value = Math.random().toString();
      expect(onChange).not.toHaveBeenCalled();
      wrapper.find('.input').simulate('change', {
        target: { value }
      });
      expect(onChange).toHaveBeenCalledWith(value, subjectId);
    });

    test('onCancel is called after click', () => {
      expect(onCancel).not.toHaveBeenCalled();
      wrapper.find('.cancel').simulate('click');
      expect(onCancel).toHaveBeenCalledWith(subjectId);
    });

    test('onSubmit is called after click', () => {
      expect(onSubmit).not.toHaveBeenCalled();
      wrapper.find('.submit').simulate('click');
      expect(onSubmit).toHaveBeenCalledWith(subjectId);
    });
  });
});

function createCmpWithValidDraft(): MockCmpData<Props> {
  const initialProps: Props = {
    subjectId: validDraft.subjectId,
    draft: validDraft,
    isValid: true,
    onChange: jest.fn(),
    onSubmit: jest.fn(),
    onCancel: jest.fn()
  };
  const wrapper = shallow<Props, {}>(
    <ReplyFormComponent {...initialProps} />
  );
  return {
    initialProps,
    wrapper
  };
}

function createCmpWithInvalidDraft(): MockCmpData<Props> {
  const initialProps: Props = {
    subjectId: invalidDraft.subjectId,
    draft: invalidDraft,
    isValid: false,
    onChange: jest.fn(),
    onSubmit: jest.fn(),
    onCancel: jest.fn()
  };
  const wrapper = shallow<Props, {}>(
    <ReplyFormComponent {...initialProps} />
  );
  return {
    initialProps,
    wrapper
  };
}
