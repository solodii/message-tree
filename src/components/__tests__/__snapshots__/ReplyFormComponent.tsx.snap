// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReplyFormComponent INVALID_DRAFT_SUBJECT renders correctly 1`] = `
<div
  className="form"
>
  <div>
    <textarea
      autoFocus={true}
      className="input"
      onChange={[Function]}
      placeholder="Type your message here..."
      value=""
    />
  </div>
  <div>
    <button
      className="cancel"
      onClick={[Function]}
    >
      Cancel
    </button>
    <button
      className="submit"
      disabled={true}
      onClick={[Function]}
    >
      Submit
    </button>
  </div>
</div>
`;

exports[`ReplyFormComponent VALID_DRAFT_SUBJECT renders correctly 1`] = `
<div
  className="form"
>
  <div>
    <textarea
      autoFocus={true}
      className="input"
      onChange={[Function]}
      placeholder="Type your message here..."
      value="draft for 0a6a7bfd-a212-4ae1-8125-6c7486213a94"
    />
  </div>
  <div>
    <button
      className="cancel"
      onClick={[Function]}
    >
      Cancel
    </button>
    <button
      className="submit"
      disabled={false}
      onClick={[Function]}
    >
      Submit
    </button>
  </div>
</div>
`;
